File: day1/day1.py
import sys

def sum_matching(line, index_func=lambda i: i + 1):
    """
    Computes the sum of matching pairs in a circular list.

    A matching pair is when a number equals the number found at index_func(i)
    in the list, where i is the numbers index and index_func is the provided
    function. By default this is the next number, i.e. i + 1.

    Parameters:
        line (str): A string of numbers.
        index_func (func): The function used to find a numbers match.

    Returns:
        (int): The sum of matching pairs.
    """
    max = len(line)
    sum = 0
    for i, num in enumerate(line):
        if num == line[index_func(i) % max]:
            sum += int(num)
    return sum

def run(stdin):
    """
    Takes problem input and yields solutions.

    Parameters:
        stdin (str): The input to the problem as a string.

    Yields:
        (*): The solutions to the problem.
    """
    line = stdin.splitlines()[0]
    max = len(line)

    yield sum_matching(line)
    yield sum_matching(line, index_func=lambda i: i + (max // 2))

if __name__ == "__main__":
    results = run(sys.stdin.read())
    for result in results:
        print(result)

File: day1/day1.c
#include <stdlib.h>
#include <stdio.h>
#include <string.h>

#include "../api.h"

int sum_matching(char* line, int max, int (*func)(int, int)) {
    int sum = 0;
    for (int i = 0; i < max; i++) {
        if (line[i] == line[func(i, max) % max]) {
			sum += (int) (line[i] - '0');
		}
    }
    return sum;
}

int addOne(int i, int max) {
    return i + 1;
}

int halfway(int i, int max) {
    return i + (max / 2);
}

int main(int argc, char** argv) {
	Input* input = load_input();

	char* string = input->lines[0];
	int length = input->sizes[0];

    printf("%d\n", sum_matching(string, length, addOne));
    printf("%d\n", sum_matching(string, length, halfway));
}
